core/src/main/scala/filodb.core/columnstore/Analyzer.scala

```diff
@@ -0,0 +1,92 @@
+package filodb.core.columnstore
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.Try
+
+import filodb.core.Types._
+
+case class Histogram(min: Int, max: Int, sum: Int, numElems: Int, buckets: Map[Int, Int]) {
+  /** Adds an element to the Histogram, adding it to histogram buckets according to the
+   *  bucket keys
+   */
+  def add(elem: Int, bucketKeys: Array[Int]): Histogram = {
+    // See the JavaDoc for binarySearch...
+    val bucket = java.util.Arrays.binarySearch(bucketKeys, elem) match {
+      case x: Int if x < 0  =>  -x - 2
+      case x: Int if x >= bucketKeys.size =>  bucketKeys.size - 1
+      case x: Int => x
+    }
+    val key = bucketKeys(bucket)
+    Histogram(Math.min(elem, this.min),
+              Math.max(elem, this.max),
+              this.sum + elem,
+              this.numElems + 1,
+              buckets + (key -> (buckets.getOrElse(key, 0) + 1)))
```

Comment:
copy(
         min = Math.min(elem, this.min),
         max = Math.max(elem, this.max),
         sum = this.sum + elem,
         numElems = this.numElems + 1,
         buckets = buckets + (key -> (buckets.getOrElse(key, 0) + 1)))